server/channels/store/sqlstore/desktop_tokens_store.go

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sqlstore
+
+import (
+	"database/sql"
+
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+	"github.com/mattermost/mattermost/server/v8/einterfaces"
+	sq "github.com/mattermost/squirrel"
+	"github.com/pkg/errors"
+)
+
+type SqlDesktopTokensStore struct {
+	*SqlStore
+	metrics einterfaces.MetricsInterface
+}
+
+func newSqlDesktopTokensStore(sqlStore *SqlStore, metrics einterfaces.MetricsInterface) store.DesktopTokensStore {
+	return &SqlDesktopTokensStore{
+		SqlStore: sqlStore,
+		metrics:  metrics,
+	}
+}
+
+func (s *SqlDesktopTokensStore) GetUserId(desktopToken, serverToken string, minCreateAt int64) (*string, error) {
+	query := s.getQueryBuilder().
+		Select("UserId").
+		From("DesktopTokens").
+		Where(sq.And{
+			sq.Eq{"DesktopToken": desktopToken},
+			sq.Eq{"ServerToken": serverToken},
+			sq.GtOrEq{"CreateAt": minCreateAt},
+		})
+
+	dt := struct{ UserId sql.NullString }{}
+	err := s.GetReplicaX().GetBuilder(&dt, query)
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, store.NewErrNotFound("DesktopTokens", desktopToken)
+		}
+		return nil, errors.Wrapf(err, "No token for %s", desktopToken)
+	}
+
+	// Check if the string is NULL, if so throw not found
+	if !dt.UserId.Valid {
+		return nil, store.NewErrNotFound("DesktopTokens", desktopToken)
+	}
+
+	return &dt.UserId.String, nil
+}
+
+func (s *SqlDesktopTokensStore) Insert(desktopToken string, createAt int64, userId *string) error {
+	builder := s.getQueryBuilder().
+		Insert("DesktopTokens").
+		Columns("DesktopToken", "CreateAt", "UserId").
+		Values(desktopToken, createAt, userId)
```

Comment:
Probably not a big deal, but `desktopToken` is a string provided through the API. Any need to validate it? squirrel should deal with any weird characters and I guess we are just making comparisons with this string, so there shouldn't be any issue, but I wanted to raise it just in case.

Reply:
Yep Squirrel produces a [prepared statement](https://en.wikipedia.org/wiki/Prepared_statement) and the RDBMS takes care of combining the query with the inputs.